wagtail/models/__init__.py

```diff
@@ -2907,20 +2908,123 @@ def __str__(self):
 
 
 class UserPagePermissionsProxy:
-    """Helper object that encapsulates all the page permission rules that this user has
-    across the page hierarchy."""
+    """
+    Helper object that encapsulates all the page permission rules
+    that this user has across the page hierarchy.
+    """
 
-    def __init__(self, user):
-        self.user = user
+    def __new__(cls, user):
+        """
+        Creates a new instance once for each user to avoid repeated queries
+        in the same request-response cycle.
+        This relies on __init__ not being implemented as it'd override the
+        initialisation we did here.
+        """
+        if not hasattr(user, "_page_permissions_proxy"):
+            self = super().__new__(cls)
+            self.user = user
 
-        if user.is_active and not user.is_superuser:
-            self.permissions = GroupPagePermission.objects.filter(
-                group__user=self.user
-            ).select_related("page")
+            # Cache this instance in the user's dict so we can reuse it.
+            user._page_permissions_proxy = self
 
-    def revisions_for_moderation(self):
-        """Return a queryset of page revisions awaiting moderation that this user has publish permission on"""
+            if not user.is_active or user.is_superuser:
+                return self
+
+            prefetch_user_perms = Prefetch(
+                "group_permissions",
+                queryset=GroupPagePermission.objects.filter(group__user=user),
+                to_attr="_perms_for_user",
+            )
+            # Load all the distinct pages the user has any permission for.
+            self._pages = (
+                Page.objects.filter(
+                    group_permissions__group__user=user,
+                )
+                .prefetch_related(prefetch_user_perms)
+                .distinct()
+                .in_bulk()
+            )
+            # Since a lot of queries involve looking up pages
+            # the user has a specific permission for, we
+            # build this mapping of `permission_type: {page_pks}`
+            # to retrieve that information quickly without asking
+            # the database.
+            self.perm_types = defaultdict(set)
+            for page_pk, page in self._pages.items():
+                # Rewrite the `_perms_for_user` annotation and only
+                # store the different type of permission the user has.
+                page._perms_for_user = {
+                    perm.permission_type for perm in page._perms_for_user
+                }
```

Comment:
Suggestion: It should be possible to do this entirely within `prefetch_related`. If all you need is the list of them, it might be faster to use `Concat` instead and split them later.